### PR TITLE
Initial implementation of multi stack alarms

### PR DESCRIPTION
**Pull request by stephenthrelfall**
Related to #207 

Seems quite straightforward to implement, doesn't yet account for the possibility of orphaned stacks but that can be addressed in a future PR

### CODE REVIEW

File: Watchman.Engine/Generation/Generic/CloudWatchCloudFormationMultiStackTemplate.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Newtonsoft.Json.Linq;
+using Watchman.Configuration;
+
+namespace Watchman.Engine.Generation.Generic
+{
+    public class CloudWatchCloudFormationMultiStackTemplate
```
Comment (tomhaigh) [inline Watchman.Engine/Generation/Generic/CloudWatchCloudFormationMultiStackTemplate.cs]: It feels a bit like the multiple-stack logic is spread across this class and `CloudFormationAlarmCreator`
I wonder if it might be better to have a different implementation of `IAlarmCreator` and try to pull any shared logic out elsewhere

Comment (tomhaigh) [reply Watchman.Engine/Generation/Generic/CloudWatchCloudFormationMultiStackTemplate.cs]: although actually maybe we don't even need the old implementation

File: Watchman.Engine.IntegrationTests/AlarmFinderTests.cs [hunk 2]
```diff
@@ -32,8 +32,7 @@ public async Task TestCount()
 
         private static IAlarmFinder MakeAlarmFinder()
         {
-            var cloudwatch = new AmazonCloudWatchClient(CredentialsReader.GetCredentials(),
-                new AmazonCloudWatchConfig {RegionEndpoint = RegionEndpoint.EUWest1});
+            var cloudwatch = new AmazonCloudWatchClient(RegionEndpoint.EUWest1);
```
Comment (tomhaigh) [inline Watchman.Engine.IntegrationTests/AlarmFinderTests.cs]: This change is ok becuase the SDK will read from the config file anyway by itself?

File: Watchman.Engine/Generation/Generic/CloudFormationAlarmCreator.cs [hunk 3]
```diff
@@ -73,20 +73,28 @@ public async Task SaveChanges(bool dryRun)
...
-                    await GenerateAndDeployStack(
-                        alarms,
-                        alertingGroup.Targets,
-                        alertingGroup.Name,
-                        stackName,
-                        dryRun);
-                }
-                catch (Exception e)
+
+                for (int i = 0; i < alertingGroup.NumberOfCloudFormationStacks; i++)
```
Comment (tomhaigh) [inline Watchman.Engine/Generation/Generic/CloudFormationAlarmCreator.cs]: minor
wonder if it might be a bit cleaner to take the alarms and GroupBy the results of Bucket()

then iterate the groups to deploy each stack. you would then have access to the group index and could remove `i` and `iClosure`

File: Watchman.Engine/Generation/Generic/CloudWatchCloudFormationMultiStackTemplate.cs (hunk 1 above)
Comment (tomhaigh) [inline Watchman.Engine/Generation/Generic/CloudWatchCloudFormationMultiStackTemplate.cs]: Is this file still needed?

**Comment by tomhaigh** [comment]
Looks good. Only minor questions above.
Might be nice to add an overall test (like the ones in watchman.tests)

File: Watchman.Engine/Generation/Generic/CloudFormationAlarmCreator.cs (hunk 3 above)
Comment (stephenthrelfall) [reply Watchman.Engine/Generation/Generic/CloudFormationAlarmCreator.cs]: good call

File: Watchman.Engine.IntegrationTests/AlarmFinderTests.cs (hunk 2 above)
Comment (stephenthrelfall) [reply Watchman.Engine.IntegrationTests/AlarmFinderTests.cs]: that's right, was causing failures on my local machine

File: Watchman.Engine/Generation/Generic/CloudFormationAlarmCreator.cs (hunk 3 above)
Comment (stephenthrelfall) [reply Watchman.Engine/Generation/Generic/CloudFormationAlarmCreator.cs]: Tried to clean this up but it failed situations where we have no alarms (thank god for end to end tests), so just left it as it was.
